### PR TITLE
fallback: Handle GetNextVariableName returning EFI_BUFFER_TOO_SMALL

### PR DESCRIPTION
**Pull request by jprvita**
Now that fallback lists all EFI variables instead of only `Boot????` variables, a 256 chars buffer may not be large enough for the variable's name. This PR changes it to a dynamic buffer that starts with 256 chars, but gets extended if needed. It also adds extra logs when `GetNextVariableName` returns an error.

This is a follow-up to #312 -- these changes were mixed with other downstream commits and were left behind.

### CODE REVIEW

**Comment by jprvita** [comment]
For-pushed rebase on latest main.

**Comment by jprvita** [comment]
Force-pushed rebase on latest main.

**Comment by jprvita** [comment]
Force-pushed rebase on latest main.

File: fallback.c [hunk 1]
```diff
@@ -431,8 +431,26 @@ find_boot_option(EFI_DEVICE_PATH *dp, EFI_DEVICE_PATH *fulldp,
 		UINTN varname_size = sizeof(varname);
 		efi_status = gRT->GetNextVariableName(&varname_size, varname,
 						      &vendor_guid);
-		if (EFI_ERROR(efi_status))
+		if (EFI_ERROR(efi_status)) {
+			if (efi_status == EFI_BUFFER_TOO_SMALL)
+				VerbosePrint(L"Buffer too small for next var "
+					     L"name\n");
```
Comment (vathpela) [inline fallback.c]: Don't break the string across lines here.

File: fallback.c [hunk 2]
```diff
@@ -431,8 +431,26 @@ find_boot_option(EFI_DEVICE_PATH *dp, EFI_DEVICE_PATH *fulldp,
...
+
+			if (efi_status == EFI_DEVICE_ERROR)
+				VerbosePrint(L"The next variable name could "
+					     L"not be retrieved due to a "
+					     L"hardware error\n");
```
Comment (vathpela) [inline fallback.c:454]: Don't break the string across lines here.

File: fallback.c [hunk 3]
```diff
@@ -431,8 +431,26 @@ find_boot_option(EFI_DEVICE_PATH *dp, EFI_DEVICE_PATH *fulldp,
...
+
+			if (efi_status == EFI_INVALID_PARAMETER)
+				VerbosePrint(L"Invalid parameter to "
+					     L"GetNextVariableName: "
+					     L"varname_size=%d, varname=%s\n",
```
Comment (vathpela) [inline fallback.c:459]: Don't break the string across lines here.

File: fallback.c [hunk 4]
```diff
@@ -426,15 +429,24 @@ find_boot_option(EFI_DEVICE_PATH *dp, EFI_DEVICE_PATH *fulldp,
...
+		UINTN varname_size = buffer_size;
 		efi_status = gRT->GetNextVariableName(&varname_size, varname,
 						      &vendor_guid);
 		if (EFI_ERROR(efi_status)) {
-			if (efi_status == EFI_BUFFER_TOO_SMALL)
+			if (efi_status == EFI_BUFFER_TOO_SMALL) {
 				VerbosePrint(L"Buffer too small for next var "
-					     L"name\n");
+					     L"name, re-allocating it to be %d"
+					     L" chars long and retrying\n",
```
Comment (vathpela) [inline fallback.c:440]: Don't break the string across lines here (even though it was incorrectly broken before.)

**Review by vathpela — CHANGES_REQUESTED** [review]
A couple of minor nits, but this basically looks good to me.

**Comment by vathpela** [comment]
I've pushed this as 437caecbab1 and 77648c5c01c .

**Pull request closed without merging**